### ISSUE TITLE
BULK create

### ISSUE THREAD

**Issue by nikitasherman** [mongoid/mongoid#4117]
Hi, It seems that in current version (master)
when I'm passing array of items to Mongoid's create method (example: Car.create([{},...{}])
it actually evaluates many inserts instead of bulk insert. I'm not sure it should work this way.
I know I can user Model.collection.insert_many...but then I loose all the Mongoid layer.

**Comment by durran** [comment]
Thanks for pointing that out - will change this.

**Comment by Shwetakale** [comment]
I think mongoid does this because all validations and callbacks should be triggered with create method and if we use bulk insert then this won't be possible.

**Comment by durran** [comment]
Yeah I forgot about that. You'll have to drop to the driver if you want to do the bulk insert because of this.

**Comment by durran** [comment]
Actually, we can add

**Comment by Shwetakale** [comment]
@durran And will that bulk_create will skip validations and callback?

**Comment by nikitasherman** [comment]
Hi, thanks for the effort to help!
@Shwetakale is right, there is the callbacks issue. I made for now for myself a kind of bulk_create
and yes without the callbacks. Which of course is not so good, because then I lose the Timestamps module which is implemented around the :create, :update callbacks. If there still any possibility to implement it with callbacks it will be great. I also would like to help as much as I can.

**Comment by durran** [comment]
This depends...

The implementation would need to run all the validations for all documents while collecting them into a single array of documents to send to the bulk write. Then we'd need to run all the "before" callbacks for each of those documents. Then we send them to the driver, and then once the response comes we'd have to do a couple things... If successful, then we run all the "after" callbacks for each document. If failed, then we need to inspect the response and see at what point it failed, and what got inserted and what didn't. The documents that did get inserted would then run only their "after" callbacks.

The complexity of doing it with validations and callbacks might cancel out the performance gain of doing it in bulk vs. individual writes. I would need to test performance on both on various data sets to see if it was truly worth it or not. If not, then I would say the `bulk_write` doesn't do it, but if it's still faster in most cases then I would say we do... Just needs a good amount of investigation.

**Comment by nikitasherman** [comment]
Sounds good to me. I hope it will not force a very big change in current callbacks implementation. Anyway thank you!
